### PR TITLE
Staging to master to add xlearn

### PR DESCRIPTION
**Pull request by miguelgfierro**
### Checklist:

- [ ] I have followed the [contribution guidelines](../CONTRIBUTING.md) and code style for this project.
- [ ] I have added tests covering my contributions.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by miguelgfierro** [comment]
I did this PR because there is a person in https://www.xbox.com/en-GB/xbox-game-streaming/project-xcloud that can be interested in xlearn. I just realized that we need to add a test for this notebook. @yueguoguo are you working on this? if not I can go ahead and add it

**Comment by yueguoguo** [comment]
@miguelgfierro ah yea we decided to create another PR for the `xlearn` notebook testing. Please do so if you have already started the development otherwise I can help. Thanks!

**Comment by miguelgfierro** [comment]
there are still some issues in the nightly builds in windows, please hold this PR until they are solved

**Comment by miguelgfierro** [comment]
`dsvm_notebook_win_pyspark` is not being executed o_o I'll follow up with this

**Comment by miguelgfierro** [comment]
> dsvm_notebook_win_pyspark is not being executed o_o I'll follow up with this

ok the test was not selected in the configuration, that was why it was not being executed.

Ready to review guys @anargyri @yueguoguo

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -12,4 +12,5 @@
 <!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
 - [ ] I have followed the [contribution guidelines](../CONTRIBUTING.md) and code style for this project.
 - [ ] I have added tests covering my contributions.
-- [ ] I have updated the documentation accordingly.
\ No newline at end of file
+- [ ] I have updated the documentation accordingly.
+- [ ] This PR is being made to `staging` and not `master`.
```
Comment (dciborow) [inline .github/PULL_REQUEST_TEMPLATE.md]: This is particularly confusing for me, and I have previously made PRs into Master, which were also mistaken accepted.

It would make things much easier if the "Default Branch" in the repo branch "Settings" was set to staging instead of "Master" to prevent this from happening with people who are expected master to be the main branch.

image

File: docker/Dockerfile [hunk 2]
```diff
@@ -14,7 +14,7 @@ WORKDIR ${HOME}
 
 # Install base dependencies
 RUN apt-get update && \
-    apt-get install -y curl git
+    apt-get install -y curl git
```
Comment (dciborow) [inline docker/Dockerfile]: extra whitespace snuck in

File: notebooks/00_quick_start/sar_movielens.ipynb [hunk 3]
```diff
@@ -115,14 +114,14 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 73,
+   "execution_count": 4,
    "metadata": {},
    "outputs": [
```
Comment (dciborow) [inline notebooks/00_quick_start/sar_movielens.ipynb]: I think it is strange to include the notebook output sections. Though the ones in this notebook look clean, they will vanish as soon as the user executes the notebook. If it were put into Markdown before or after the code execution, then the user can see that their output is the same as the expected output.

Comment (dciborow) [reply notebooks/00_quick_start/sar_movielens.ipynb]: The updated outputs also make it a bit hard to follow the delta with this notebook in the code review

File: scripts/generate_conda_file.py [hunk 4]
```diff
@@ -55,14 +55,15 @@
     "swig": "swig==3.0.12",
     "tensorflow": "tensorflow==1.12.0",
     "lightgbm": "lightgbm==2.2.1",
+    "cmake": "cmake==3.14.0",
```
Comment (dciborow) [inline scripts/generate_conda_file.py]: is it typical for CONDA to have the libraries in pseudorandom order compared to alphabetical?

File: tests/ci/azure_pipeline_test/dsvm_nightly_linux_cpu.yml [hunk 5]
```diff
@@ -31,18 +31,16 @@ jobs:
...
       conda env create --quiet -f nightly_reco_base.yaml 2> log
     displayName: 'Setup Conda Env'
-    timeoutInMinutes: 10
 
   - script: |
       . /anaconda/etc/profile.d/conda.sh && \
       conda activate nightly_reco_base && \
       echo "Smoke tests" && \
-      pytest tests/smoke -m "smoke and not spark and not gpu" --junitxml=reports/test-smoke.xml && \
+      pytest tests/smoke --durations 0 -m "smoke and not spark and not gpu" --junitxml=reports/test-smoke.xml && \
```
Comment (dciborow) [inline tests/ci/azure_pipeline_test/dsvm_nightly_linux_cpu.yml]: when using the "script: |" action, you do not need to include && \ on every line, and can instead remove all of these (unless they are specifically desired).
This seems like legacy from pre-yml ADO

File: tests/ci/azure_pipeline_test/dsvm_nightly_linux_cpu.yml [hunk 6]
```diff
@@ -31,18 +31,16 @@ jobs:
       python ./scripts/generate_conda_file.py --name nightly_reco_base && \
       conda env create --quiet -f nightly_reco_base.yaml 2> log
     displayName: 'Setup Conda Env'
-    timeoutInMinutes: 10
 
   - script: |
```
Comment (dciborow) [inline tests/ci/azure_pipeline_test/dsvm_nightly_linux_cpu.yml]: This task should include a "failOnStandardError: True" so that if the conda install does not work, the task will fail and the build will not continue.

File: tests/ci/azure_pipeline_test/dsvm_nightly_win_cpu.yml [hunk 7]
```diff
@@ -12,51 +12,49 @@ trigger: none
...
\ No newline at end of file
+jobs:
+- job: nightly
+  displayName: 'Nightly tests Windows CPU'
+  timeoutInMinutes: 180 # how long to run the job before automatically cancelling
+  pool:
+    name: RecommendersAgentPoolWin
+
+  steps:
+  - script: |
```
Comment (dciborow) [inline tests/ci/azure_pipeline_test/dsvm_nightly_win_cpu.yml]: include failOnStandardError: True

File: tests/ci/azure_pipeline_test/dsvm_nightly_win_pyspark.yml [hunk 8]
```diff
@@ -12,48 +12,48 @@ trigger: none
...
-- script: |
-   call conda env remove -n nightly_reco_pyspark -y
-   rmdir /s /q C:\Anaconda\envs\nightly_reco_pyspark
-   
-  workingDirectory: tests
-  displayName: 'Conda remove'
-  continueOnError: true
-  condition: succeededOrFailed()
-  timeoutInMinutes: 10
+jobs:
```
Comment (dciborow) [inline tests/ci/azure_pipeline_test/dsvm_nightly_win_pyspark.yml]: the nightly, v. on PR/Commit builds should be templates with only input parameter differences

File: tests/ci/azure_pipeline_test/dsvm_notebook_linux_cpu.yml [hunk 9]
```diff
@@ -11,40 +11,43 @@ trigger:
...
-    testResultsFiles: '**/test-*.xml'
-    failTaskOnFailedTests: true
-  condition: succeededOrFailed()
-
+jobs:
+- job: unit
+  displayName: 'Unit tests Linux CPU on notebooks'
+  timeoutInMinutes: 20 # how long to run the job before automatically cancelling
+  pool:
+    name: recolinuxpool
```
Comment (dciborow) [inline tests/ci/azure_pipeline_test/dsvm_notebook_linux_cpu.yml]: dehard code

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 1 above)
Comment (miguelgfierro) [reply .github/PULL_REQUEST_TEMPLATE.md]: the idea is to have staging as developing branch while master is our stable branch, most users will just download master, where (hopefully) the code works always. In staging code can fail

File: notebooks/00_quick_start/sar_movielens.ipynb (hunk 3 above)
Comment (miguelgfierro) [reply notebooks/00_quick_start/sar_movielens.ipynb]: the reason for having the outputs in the notebooks is because we want the notebooks to show the results. I can go to NCF notebook on github and see very quickly how it works, if we show no outputs then there is no much difference with a python file

File: scripts/generate_conda_file.py (hunk 4 above)
Comment (miguelgfierro) [reply scripts/generate_conda_file.py]: yeah, here we just put a random order, no specific reason, but probably an alphabetical order would be nicer

File: tests/ci/azure_pipeline_test/dsvm_nightly_win_cpu.yml [hunk 10]
```diff
@@ -12,51 +12,49 @@ trigger: none
...
-  timeoutInMinutes: 10
-
-- script: |
-   del /q /S %LOCALAPPDATA%\Temp\*
-   for /d %%i in (%LOCALAPPDATA%\Temp\*) do @rmdir /s /q "%%i"
-  displayName: 'Remove Temp Files'
-  condition: succeededOrFailed()
\ No newline at end of file
+jobs:
+- job: nightly
```
[the remark this answers is not in the document]
Comment (miguelgfierro) [reply tests/ci/azure_pipeline_test/dsvm_nightly_win_cpu.yml]: this is true, in fact we sent a feedback item a couple of months ago about this. I think Dan G and Praneet had the same problem. Last time I checked, the scheduler in the yml was not working, so we had to set the scheduler from the dashboard, do you know if it is working now?

[the remark this answers is not in the document]
Comment (dciborow) [reply tests/ci/azure_pipeline_test/dsvm_nightly_win_cpu.yml]: It is still not working. But I do not believe I have seen a Feedback item on this in triage.

File: notebooks/02_model/fm_deep_dive.ipynb [hunk 11]
```diff
@@ -0,0 +1,894 @@
...
+   "metadata": {},
+   "source": [
+    "FM is an algorithm that uses factorization in prediction tasks with data set of high sparsity. The algorithm was original proposed in [\\[1\\]](https://www.csie.ntu.edu.tw/~b97053/paper/Rendle2010FM.pdf). Traditionally, the algorithms such as SVM failed in dealing with highly sparse data that is usually seen in many contemporary problems, e.g., click-through rate prediction, recommendation, etc. FM handles the problem by modeling not just first-order linear components for predicting the label, but also the cross-product of the feature variables in order to capture more generalized correlation between variables and label. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "For example, usually, the data that appears in recommendation problems can be encoded in the following way - the user, item, and feature vectors are transformed into one-hot representation for building model. For this arrangement, using the classic algorithms like linear regression, SVM, for predicting the target, are that\n",
```
Comment (loomlike) [inline notebooks/02_model/fm_deep_dive.ipynb]: This part can be improved grammatically.

File: notebooks/02_model/fm_deep_dive.ipynb [hunk 12]
```diff
@@ -0,0 +1,894 @@
...
+   "metadata": {},
+   "source": [
+    "$$<\\textbf{v}_{i}, \\textbf{v}_{j}>=\\sum^{k}_{f=1}v_{i,f}\\cdot v_{j,f}$$"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Compared to using fixed parameter for the high-order interaction components, using the factorized vectors increase generalization as well as expressiveness of the model. In addition to this, the computation complexity of the model is $O(kn)$ where $k$ and $n$ are the dimensionalities of the factorization vector and input feature vector, respectively. In practice, usually a two-way FM model is used, i.e., only the second-order feature interactions are considered, to favor computational efficiency without loss of model performance."
```
Comment (loomlike) [inline notebooks/02_model/fm_deep_dive.ipynb]: > ... computation complexity of the model is ...

To be precise, model's "prediction" complexity.

File: notebooks/02_model/fm_deep_dive.ipynb [hunk 13]
```diff
@@ -0,0 +1,894 @@
...
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "There are in total 4 fields and 10 features.\n"
+     ]
+    }
+   ],
+   "source": [
+    "print('There are in total {0} fields and {1} features.'.format(converter.field_count, converter.feature_count))"
```
Comment (loomlike) [inline notebooks/02_model/fm_deep_dive.ipynb]: In general, would be nice to have black-like formatting (using blackmagic maybe). But formatting is not super important so it is just a suggestion.

File: notebooks/02_model/fm_deep_dive.ipynb [hunk 14]
```diff
@@ -0,0 +1,894 @@
...
...WGtX4HyQ0Rc8uQ/ffYE/dKeLXQSwPdVyFNAkszrW2kRjzAGcoa0InE9zpt42wrtdldPIybnUB+HOL9k5l76TFw9P65jbfXrMZFB26r2umdXJzraSe3JyLuX8ovOUgbx4PPKF+zDpjGX3uSyVAYwx+YDiOMM1Z/JcEPFeTs6lnF/0s5UBzwLfGFPHGDPYGPOeMeZdd76utfZFr/rMI0uB2saY6saYzJ7tMRXncQgAdwA/WOcTb1OBru6dH9Vxnu+yJJf2W9LLybmU80t2zqXvePXBq8E4b5IYnABb6s6Pz+j2qAuZtTb1c1nWAl9Za1cbY142xnR2q30ClDbGbAAG4DwsDWvtauArnOe1zAT6nLiV1RgzHufBTwFjTJQxpmduHpcf5eRcAhhjtgBvAw+65+zUu0Ikl2TnXBpjGhljooA7gY+MMRf9FxV49SyddcBl1tqEU8rzA6uttbXPeaciIpIlr4Z0koGKGZRXwEe3QImInE+8ejxyP2CeMWY9J2+NqoLzfPAnPOpTRESy4NnTMo0xITgffojAGb+PApamftyCiIjkHs8fjywiIueHvLgPX0RE8oACX3zDGBPvQZtbjDFl8qJvkTOlwBcR8Qmv7tIRuSAYYzoBzwP5cb6kvJu1dpcxZhhQHedW4kicD1ldg/O43WigU6rPmQwyxrR05++11m5wPzn9Jc7P2MxU/RUBpgAlgTDgeWvtFG+PUsShK3zxu0XANdba+jifDn861bqaQAecx+r+G5hvra0H/M8tP+GgtbYx8AHwjlv2LvChtbYRsDNV3aPAbdbaBkBLYKT7qGwRzynwxe8qAbOMMSuBQThfRnPCDPcqfiUQyskr9ZVAtVT1xqf691p3vmmq8s9T1TXAa8aYFcBcnNuWy5+TIxE5DQW++N37wAfulXtvoGCqdccArLXJQEKqh6Qlk3Y41GZj/oRuQFmgobX2KmDXKX2KeEaBL35XHGdMHk4+BfNM3Z3q31/c+Z9wntAITsin7i/WWpvgjvtXPcs+Rc6Y3rQVPynkPh3xhLdxvnrzP8aYaJxvH6t+Fu0WMMb8inMBdY9b1hf40hjTF/gmVd0vgGnGmN+AP4C/zqI/kbOiT9qKiPiEhnRERHxCgS8i4hMKfBERn1Dgi4j4hAJfRMQnFPgiIj6hwBcR8Yn/B5WB+yFi76KKAAAAAElFTkSuQmCC\n",
```
Comment (loomlike) [inline notebooks/02_model/fm_deep_dive.ipynb]: matplotlib had a bug in plotting matrix like this. [Here is the related issue](https://github.com/matplotlib/matplotlib/issues/14897)

File: notebooks/02_model/fm_deep_dive.ipynb [hunk 15]
```diff
@@ -0,0 +1,894 @@
...
+    "ffm_model.setValidate(valid_file)  # Set the path of validation dataset\n",
+    "\n",
+    "# Parameters:\n",
+    "#  0. task: binary classification\n",
+    "#  1. learning rate: 0.2\n",
+    "#  2. regular lambda: 0.002\n",
+    "#  3. evaluation metric: auc\n",
+    "#  4. number of epochs: 10\n",
+    "#  5. optimization method: sgd\n",
+    "param = {\"task\":\"binary\", \n",
```
Comment (loomlike) [inline notebooks/02_model/fm_deep_dive.ipynb]: Same here, about formatting... but again, not super important. No pressure. ;-)

Comment (miguelgfierro) [reply notebooks/02_model/fm_deep_dive.ipynb]: added issue to work on it when we have time https://github.com/microsoft/recommenders/issues/1000

File: notebooks/02_model/fm_deep_dive.ipynb (hunk 13 above)
Comment (miguelgfierro) [reply notebooks/02_model/fm_deep_dive.ipynb]: added issue to work on it when we have time https://github.com/microsoft/recommenders/issues/1000

File: notebooks/02_model/fm_deep_dive.ipynb (hunk 12 above)
Comment (miguelgfierro) [reply notebooks/02_model/fm_deep_dive.ipynb]: @yoeguoguo can you comment on this?

File: notebooks/02_model/fm_deep_dive.ipynb (hunk 11 above)
Comment (miguelgfierro) [reply notebooks/02_model/fm_deep_dive.ipynb]: please see https://github.com/microsoft/recommenders/pull/1001

File: notebooks/02_model/fm_deep_dive.ipynb (hunk 14 above)
Comment (miguelgfierro) [reply notebooks/02_model/fm_deep_dive.ipynb]: right now we have `matplotlib>=2.2.2`. Jun, do you think that if we set `matplotlib>=3.1.1` and rerun the notebook, the behavior would be as expected? 
FYI @yueguoguo

Comment (loomlike) [reply notebooks/02_model/fm_deep_dive.ipynb]: My comment is just to aware of the issue, but I would suggest to stick with what we have now (anyway we use `>=`) because, 
we had the same issue at CVBP (confusion matrix y-axis was off), and when we tried to force to use the latest one, something was broken from somewhere...I can't remember exactly... maybe one of sklearn, numpy, pandas or fastai.

File: notebooks/02_model/fm_deep_dive.ipynb (hunk 12 above)
Comment (yueguoguo) [reply notebooks/02_model/fm_deep_dive.ipynb]: Yea it can be made more clear that the computation of the pairwise interaction between the vectors and interacting features (see [here](https://www.csie.ntu.edu.tw/~b97053/paper/Rendle2010FM.pdf) is linear.

**Comment by anargyri** [comment]
Some comments on fm_deep_dive.ipynb.
- In 1.1. it sounds like SVM, LR etc. cannot accept sparse features. This is not the case and I would change the phrasing to avoid giving this impression to the reader.
- I would not say there is no loss of model performance when increasing the order of FM to > 2, because we don't know what happens.
- In 2.1 I don't agree that VW is less flexible or configurable than other methods. I think it has similar parameters and functionality.
- In 2.3 I don't understand why the effect of k is small; FFM allows for user-specified k in the algorithm (both in the paper and the GitHub library). Is this based on empirical observation?

**Comment by miguelgfierro** [comment]
@yueguoguo do you have time this week to review Andreas' and Jun's comments?

File: docker/Dockerfile [hunk 16]
```diff
@@ -43,6 +43,9 @@ RUN mkdir ${HOME}/.jupyter && \
 FROM base AS cpu
 
 # Setup Conda environment
+RUN apt-get update && \
+    apt-get install -y python-pip && \
+    pip install cmake
```
Comment (gramhagen) [inline docker/Dockerfile]: why is this needed? isn't it handled within conda?

Comment (miguelgfierro) [reply docker/Dockerfile]: that's a good point, I think it is not necessary since it was added in staging recently https://github.com/microsoft/recommenders/blob/staging/scripts/generate_conda_file.py#L58
FYI @yueguoguo

Comment (yueguoguo) [reply docker/Dockerfile]: Good catch. It should have been removed.

Comment (yueguoguo) [reply docker/Dockerfile]: Okay tested it. The conda file takes care of `cmake` and it installs the package, so explicit installation is not needed. Let me remove it.

**Comment by anargyri** [comment]
There is still the sentence (in the table in 2.3) about the effect of k being small. I would remove it since it contradicts our past experience with factorization methods.

**Comment by yueguoguo** [comment]
> There is still the sentence (in the table in 2.3) about the effect of k being small. I would remove it since it contradicts our past experience with factorization methods.

This is from the original [paper](https://www.csie.ntu.edu.tw/~cjlin/papers/ffm.pdf), Section 4.2.

**Comment by anargyri** [comment]
> > There is still the sentence (in the table in 2.3) about the effect of k being small. I would remove it since it contradicts our past experience with factorization methods.
> 
> This is from the original [paper](https://www.csie.ntu.edu.tw/~cjlin/papers/ffm.pdf), Section 4.2.

Ok, let's leave it then. It may hold for FFM with enough fields, but it doesn't apply to FM etc.
